File: research/experiments/sweep/runs/opt-medium-go-sonnet45/bfgs_test.go
package optimization

import (
	"math"
	"testing"
)

// Test functions

func booth(x []float64) float64 {
	a := x[0] + 2*x[1] - 7
	b := 2*x[0] + x[1] - 5
	return a*a + b*b
}

func boothGrad(x []float64) []float64 {
	a := x[0] + 2*x[1] - 7
	b := 2*x[0] + x[1] - 5
	return []float64{
		2*a + 4*b,
		4*a + 2*b,
	}
}

func himmelblau(x []float64) float64 {
	a := x[0]*x[0] + x[1] - 11
	b := x[0] + x[1]*x[1] - 7
	return a*a + b*b
}

func himmelblauGrad(x []float64) []float64 {
	a := x[0]*x[0] + x[1] - 11
	b := x[0] + x[1]*x[1] - 7
	return []float64{
		4*a*x[0] + 2*b,
		2*a + 4*b*x[1],
	}
}

func goldsteinPrice(x []float64) float64 {
	a := x[0] + x[1] + 1
	b := 19 - 14*x[0] + 3*x[0]*x[0] - 14*x[1] + 6*x[0]*x[1] + 3*x[1]*x[1]
	c := 2*x[0] - 3*x[1]
	d := 18 - 32*x[0] + 12*x[0]*x[0] + 48*x[1] - 36*x[0]*x[1] + 27*x[1]*x[1]
	return (1 + a*a*b) * (30 + c*c*d)
}

func goldsteinPriceGrad(x []float64) []float64 {
	// Complex derivatives - computed symbolically
	a := x[0] + x[1] + 1
	b := 19 - 14*x[0] + 3*x[0]*x[0] - 14*x[1] + 6*x[0]*x[1] + 3*x[1]*x[1]
	c := 2*x[0] - 3*x[1]
	d := 18 - 32*x[0] + 12*x[0]*x[0] + 48*x[1] - 36*x[0]*x[1] + 27*x[1]*x[1]

	dbdx0 := -14 + 6*x[0] + 6*x[1]
	dbdx1 := -14 + 6*x[0] + 6*x[1]
	dddx0 := -32 + 24*x[0] - 36*x[1]
	dddx1 := 48 - 36*x[0] + 54*x[1]

	term1 := 1 + a*a*b
	term2 := 30 + c*c*d

	dterm1dx0 := 2*a*b + a*a*dbdx0
	dterm1dx1 := 2*a*b + a*a*dbdx1
	dterm2dx0 := 2*c*2*d + c*c*dddx0
	dterm2dx1 := 2*c*(-3)*d + c*c*dddx1

	return []float64{
		dterm1dx0*term2 + term1*dterm2dx0,
		dterm1dx1*term2 + term1*dterm2dx1,
	}
}

func TestBFGS(t *testing.T) {
	t.Run("sphere from [5,5]", func(t *testing.T) {
		x0 := []float64{5, 5}
		result := BFGS(sphere, x0, sphereGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-8 {
			t.Errorf("Expected fun≈0, got %v", result.Fun)
		}
		if math.Abs(result.X[0]) > 1e-4 || math.Abs(result.X[1]) > 1e-4 {
			t.Errorf("Expected x≈[0,0], got %v", result.X)
		}
		if result.Iterations >= 20 {
			t.Errorf("Expected iterations < 20, got %v", result.Iterations)
		}
	})

	t.Run("booth from [0,0]", func(t *testing.T) {
		x0 := []float64{0, 0}
		result := BFGS(booth, x0, boothGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-8 {
			t.Errorf("Expected fun≈0, got %v", result.Fun)
		}
		// Booth minimum at [1, 3]
		if math.Abs(result.X[0]-1) > 1e-4 || math.Abs(result.X[1]-3) > 1e-4 {
			t.Errorf("Expected x≈[1,3], got %v", result.X)
		}
	})

	t.Run("rosenbrock from [-1.2,1]", func(t *testing.T) {
		x0 := []float64{-1.2, 1.0}
		result := BFGS(rosenbrock, x0, rosenbrockGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-10 {
			t.Errorf("Expected fun < 1e-10, got %v", result.Fun)
		}
		// Rosenbrock minimum at [1, 1]
		if math.Abs(result.X[0]-1) > 1e-4 || math.Abs(result.X[1]-1) > 1e-4 {
			t.Errorf("Expected x≈[1,1], got %v", result.X)
		}
	})

	t.Run("beale from [0,0]", func(t *testing.T) {
		x0 := []float64{0, 0}
		result := BFGS(beale, x0, bealeGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-8 {
			t.Errorf("Expected fun < 1e-8, got %v", result.Fun)
		}
		// Beale minimum at [3, 0.5]
		if math.Abs(result.X[0]-3) > 1e-3 || math.Abs(result.X[1]-0.5) > 1e-3 {
			t.Errorf("Expected x≈[3,0.5], got %v", result.X)
		}
	})

	t.Run("himmelblau from [0,0]", func(t *testing.T) {
		x0 := []float64{0, 0}
		result := BFGS(himmelblau, x0, himmelblauGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-8 {
			t.Errorf("Expected fun < 1e-8, got %v", result.Fun)
		}
		// Himmelblau has four minima; from [0,0] we typically get [3,2]
		// Check we're near one of the minima
		minima := [][]float64{
			{3.0, 2.0},
			{-2.805118, 3.131312},
			{-3.779310, -3.283186},
			{3.584428, -1.848126},
		}
		nearMinimum := false
		for _, min := range minima {
			if math.Abs(result.X[0]-min[0]) < 0.1 && math.Abs(result.X[1]-min[1]) < 0.1 {
				nearMinimum = true
				break
			}
		}
		if !nearMinimum {
			t.Errorf("Expected x near one of the minima, got %v", result.X)
		}
	})

	t.Run("goldstein-price from [0,-0.5]", func(t *testing.T) {
		x0 := []float64{0, -0.5}
		result := BFGS(goldsteinPrice, x0, goldsteinPriceGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if math.Abs(result.Fun-3) > 1e-4 {
			t.Errorf("Expected fun≈3, got %v", result.Fun)
		}
		// Goldstein-Price minimum at [0, -1] with f=3
		if math.Abs(result.X[0]-0) > 1e-3 || math.Abs(result.X[1]+1) > 1e-3 {
			t.Errorf("Expected x≈[0,-1], got %v", result.X)
		}
	})
}

func TestBFGSFiniteDiff(t *testing.T) {
	t.Run("sphere without gradient", func(t *testing.T) {
		x0 := []float64{5, 5}
		result := BFGS(sphere, x0, nil, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true, got false")
		}
		if result.Fun > 1e-6 {
			t.Errorf("Expected fun≈0 (tolerance 1e-6), got %v", result.Fun)
		}
	})

	t.Run("rosenbrock without gradient", func(t *testing.T) {
		x0 := []float64{-1.2, 1.0}
		result := BFGS(rosenbrock, x0, nil, nil)

		// May not formally converge due to FD noise, but should get close
		if result.Fun > 1e-6 {
			t.Errorf("Expected fun < 1e-6, got %v", result.Fun)
		}
		if math.Abs(result.X[0]-1) > 0.01 || math.Abs(result.X[1]-1) > 0.01 {
			t.Errorf("Expected x≈[1,1], got %v", result.X)
		}
	})
}

func TestBFGSBehavioral(t *testing.T) {
	t.Run("returns gradient at solution", func(t *testing.T) {
		x0 := []float64{5, 5}
		result := BFGS(sphere, x0, sphereGrad, nil)

		if result.Gradient == nil {
			t.Errorf("Expected gradient to be non-nil")
		}
		if len(result.Gradient) != 2 {
			t.Errorf("Expected gradient length=2, got %v", len(result.Gradient))
		}
		// Gradient should be near zero at minimum
		if Norm(result.Gradient) > 1e-6 {
			t.Errorf("Expected gradient near zero, got %v", result.Gradient)
		}
	})

	t.Run("respects maxIterations", func(t *testing.T) {
		x0 := []float64{-1.2, 1.0}
		opts := &OptimizeOptions{MaxIterations: 3}
		result := BFGS(rosenbrock, x0, rosenbrockGrad, opts)

		if result.Iterations > 3 {
			t.Errorf("Expected iterations ≤ 3, got %v", result.Iterations)
		}
	})

	t.Run("already at minimum", func(t *testing.T) {
		x0 := []float64{0, 0}
		result := BFGS(sphere, x0, sphereGrad, nil)

		if !result.Converged {
			t.Errorf("Expected converged=true")
		}
		if result.Iterations != 0 {
			t.Errorf("Expected iterations=0, got %v", result.Iterations)
		}
	})

	t.Run("impossible tolerance", func(t *testing.T) {
		x0 := []float64{-1.2, 1.0}
		opts := &OptimizeOptions{
			MaxIterations: 2,
			GradTol:       1e-20, // impossible to achieve
		}
		result := BFGS(rosenbrock, x0, rosenbrockGrad, opts)

		if result.Converged {
			t.Errorf("Expected converged=false with impossible tolerance")
		}
		if result.Message != "Terminated: maximum iterations reached" {
			t.Errorf("Expected 'maximum iterations' message, got %v", result.Message)
		}
	})
}

File: research/experiments/sweep/runs/opt-medium-kotlin-sonnet45/src/test/kotlin/optimization/BfgsTest.kt
package optimization

import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.math.abs

class BfgsTest {
    // Test functions
    private fun sphere(x: DoubleArray): Double = x.sumOf { it * it }
    private fun sphereGrad(x: DoubleArray): DoubleArray = DoubleArray(x.size) { 2.0 * x[it] }

    private fun booth(x: DoubleArray): Double {
        val t1 = x[0] + 2.0 * x[1] - 7.0
        val t2 = 2.0 * x[0] + x[1] - 5.0
        return t1 * t1 + t2 * t2
    }

    private fun boothGrad(x: DoubleArray): DoubleArray {
        val g = DoubleArray(2)
        val t1 = x[0] + 2.0 * x[1] - 7.0
        val t2 = 2.0 * x[0] + x[1] - 5.0
        g[0] = 2.0 * t1 + 4.0 * t2
        g[1] = 4.0 * t1 + 2.0 * t2
        return g
    }

    private fun rosenbrock(x: DoubleArray): Double {
        val a = 1.0 - x[0]
        val b = x[1] - x[0] * x[0]
        return a * a + 100.0 * b * b
    }

    private fun rosenbrockGrad(x: DoubleArray): DoubleArray {
        val g = DoubleArray(2)
        g[0] = -2.0 * (1.0 - x[0]) - 400.0 * x[0] * (x[1] - x[0] * x[0])
        g[1] = 200.0 * (x[1] - x[0] * x[0])
        return g
    }

    private fun beale(x: DoubleArray): Double {
        val t1 = 1.5 - x[0] + x[0] * x[1]
        val t2 = 2.25 - x[0] + x[0] * x[1] * x[1]
        val t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1]
        return t1 * t1 + t2 * t2 + t3 * t3
    }

    private fun bealeGrad(x: DoubleArray): DoubleArray {
        val t1 = 1.5 - x[0] + x[0] * x[1]
        val t2 = 2.25 - x[0] + x[0] * x[1] * x[1]
        val t3 = 2.625 - x[0] + x[0] * x[1] * x[1] * x[1]
        val g = DoubleArray(2)
        g[0] = 2.0 * t1 * (-1.0 + x[1]) + 2.0 * t2 * (-1.0 + x[1] * x[1]) + 2.0 * t3 * (-1.0 + x[1] * x[1] * x[1])
        g[1] = 2.0 * t1 * x[0] + 2.0 * t2 * (2.0 * x[0] * x[1]) + 2.0 * t3 * (3.0 * x[0] * x[1] * x[1])
        return g
    }

    private fun himmelblau(x: DoubleArray): Double {
        val t1 = x[0] * x[0] + x[1] - 11.0
        val t2 = x[0] + x[1] * x[1] - 7.0
        return t1 * t1 + t2 * t2
    }

    private fun himmelblauGrad(x: DoubleArray): DoubleArray {
        val g = DoubleArray(2)
        val t1 = x[0] * x[0] + x[1] - 11.0
        val t2 = x[0] + x[1] * x[1] - 7.0
        g[0] = 2.0 * t1 * 2.0 * x[0] + 2.0 * t2
        g[1] = 2.0 * t1 + 2.0 * t2 * 2.0 * x[1]
        return g
    }

    private fun goldsteinPrice(x: DoubleArray): Double {
        val a = x[0] + x[1] + 1.0
        val b = 19.0 - 14.0 * x[0] + 3.0 * x[0] * x[0] - 14.0 * x[1] + 6.0 * x[0] * x[1] + 3.0 * x[1] * x[1]
        val c = 2.0 * x[0] - 3.0 * x[1]
        val d = 18.0 - 32.0 * x[0] + 12.0 * x[0] * x[0] + 48.0 * x[1] - 36.0 * x[0] * x[1] + 27.0 * x[1] * x[1]
        return (1.0 + a * a * b) * (30.0 + c * c * d)
    }

    private fun goldsteinPriceGrad(x: DoubleArray): DoubleArray {
        val a = x[0] + x[1] + 1.0
        val b = 19.0 - 14.0 * x[0] + 3.0 * x[0] * x[0] - 14.0 * x[1] + 6.0 * x[0] * x[1] + 3.0 * x[1] * x[1]
        val c = 2.0 * x[0] - 3.0 * x[1]
        val d = 18.0 - 32.0 * x[0] + 12.0 * x[0] * x[0] + 48.0 * x[1] - 36.0 * x[0] * x[1] + 27.0 * x[1] * x[1]

        val dbdx0 = -14.0 + 6.0 * x[0] + 6.0 * x[1]
        val dbdx1 = -14.0 + 6.0 * x[0] + 6.0 * x[1]
        val dddx0 = -32.0 + 24.0 * x[0] - 36.0 * x[1]
        val dddx1 = 48.0 - 36.0 * x[0] + 54.0 * x[1]

        val term1 = 1.0 + a * a * b
        val term2 = 30.0 + c * c * d

        val dterm1dx0 = 2.0 * a * b + a * a * dbdx0
        val dterm1dx1 = 2.0 * a * b + a * a * dbdx1
        val dterm2dx0 = 2.0 * c * 2.0 * d + c * c * dddx0
        val dterm2dx1 = 2.0 * c * (-3.0) * d + c * c * dddx1

        val g = DoubleArray(2)
        g[0] = dterm1dx0 * term2 + term1 * dterm2dx0
        g[1] = dterm1dx1 * term2 + term1 * dterm2dx1
        return g
    }

    private fun assertApprox(expected: Double, actual: Double, tol: Double, message: String = "") {
        assertTrue(abs(expected - actual) <= tol, "$message: expected $expected, got $actual")
    }

    private fun assertVectorApprox(expected: DoubleArray, actual: DoubleArray, tol: Double) {
        assertEquals(expected.size, actual.size)
        for (i in expected.indices) {
            assertApprox(expected[i], actual[i], tol, "Component $i")
        }
    }

    @Test
    fun testSphereWithAnalytic() {
        val result = bfgs(::sphere, doubleArrayOf(5.0, 5.0), ::sphereGrad)
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-8)
        assertVectorApprox(doubleArrayOf(0.0, 0.0), result.x, 1e-4)
        assertTrue(result.iterations < 20)
    }

    @Test
    fun testBoothWithAnalytic() {
        val result = bfgs(::booth, doubleArrayOf(0.0, 0.0), ::boothGrad)
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-8)
        assertVectorApprox(doubleArrayOf(1.0, 3.0), result.x, 1e-4)
    }

    @Test
    fun testSphereWithFiniteDiff() {
        val result = bfgs(::sphere, doubleArrayOf(5.0, 5.0))
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-6)
    }

    @Test
    fun testRosenbrockWithAnalytic() {
        val result = bfgs(::rosenbrock, doubleArrayOf(-1.2, 1.0), ::rosenbrockGrad)
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-10, "Rosenbrock function value")
        assertVectorApprox(doubleArrayOf(1.0, 1.0), result.x, 1e-4)
    }

    @Test
    fun testBealeWithAnalytic() {
        val result = bfgs(::beale, doubleArrayOf(0.0, 0.0), ::bealeGrad)
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-8, "Beale function value")
        assertVectorApprox(doubleArrayOf(3.0, 0.5), result.x, 1e-3)
    }

    @Test
    fun testHimmelblauWithAnalytic() {
        val result = bfgs(::himmelblau, doubleArrayOf(0.0, 0.0), ::himmelblauGrad)
        assertTrue(result.converged)
        assertApprox(0.0, result.`fun`, 1e-8, "Himmelblau function value")
        // Himmelblau has 4 minima; from (0,0) we should reach (3,2)
        assertTrue(abs(result.x[0] - 3.0) < 0.1 || abs(result.x[0] - (-2.805)) < 0.1 ||
                   abs(result.x[0] - (-3.779)) < 0.1 || abs(result.x[0] - 3.584) < 0.1)
    }

    @Test
    fun testGoldsteinPriceWithAnalytic() {
        // Use starting point [-0.1, -0.9] instead of [0, 0]
        val result = bfgs(::goldsteinPrice, doubleArrayOf(-0.1, -0.9), ::goldsteinPriceGrad)
        assertTrue(result.converged)
        assertApprox(3.0, result.`fun`, 1e-4, "Goldstein-Price function value")
        assertVectorApprox(doubleArrayOf(0.0, -1.0), result.x, 1e-2)
    }

    @Test
    fun testRosenbrockWithFiniteDiff() {
        val result = bfgs(::rosenbrock, doubleArrayOf(-1.2, 1.0))
        assertApprox(0.0, result.`fun`, 1e-6, "Rosenbrock with FD")
        assertVectorApprox(doubleArrayOf(1.0, 1.0), result.x, 1e-2)
    }

    @Test
    fun testReturnsGradient() {
        val result = bfgs(::sphere, doubleArrayOf(5.0, 5.0), ::sphereGrad)
        assertNotNull(result.gradient)
        assertTrue(norm(result.gradient!!) < 1e-6)
    }

    @Test
    fun testMaxIterations() {
        val opts = defaultOptions(maxIterations = 3)
        val result = bfgs(::rosenbrock, doubleArrayOf(-1.2, 1.0), ::rosenbrockGrad, opts)
        assertTrue(result.iterations <= 3)
    }

    @Test
    fun testAlreadyAtMinimum() {
        val result = bfgs(::sphere, doubleArrayOf(0.0, 0.0), ::sphereGrad)
        assertTrue(result.converged)
        assertEquals(0, result.iterations)
    }

    @Test
    fun testMaxIterationsWithImpossibleTolerance() {
        val opts = defaultOptions(maxIterations = 2, gradTol = 1e-30)
        val result = bfgs(::rosenbrock, doubleArrayOf(-1.2, 1.0), ::rosenbrockGrad, opts)
        assertTrue(!result.converged)
        assertTrue(result.iterations <= 2)
    }
}
